net/net.go

```diff
@@ -127,48 +131,71 @@ func (t *net) Close() (err error) {
 		}
 	}
 
-	weakClose("DAGService", t.DAGService)
-	weakClose("host", t.host)
-	weakClose("threadstore", t.store)
+	weakClose("DAGService", n.DAGService)
+	weakClose("host", n.host)
+	weakClose("threadstore", n.store)
 
-	t.bus.Discard()
-	t.cancel()
+	n.bus.Discard()
+	n.cancel()
 
 	if len(errs) > 0 {
 		return fmt.Errorf("failed while closing net; err(s): %q", errs)
 	}
 
-	t.pullLock.Lock()
-	defer t.pullLock.Unlock()
+	n.pullLock.Lock()
+	defer n.pullLock.Unlock()
 	// Wait for all thread pulls to finish
-	for _, semaph := range t.pullLocks {
+	for _, semaph := range n.pullLocks {
 		semaph <- struct{}{}
 	}
 	return nil
 }
 
-func (t *net) Host() host.Host {
-	return t.host
+func (n *net) Host() host.Host {
+	return n.host
 }
 
-func (t *net) Store() lstore.Logstore {
-	return t.store
+func (n *net) Store() lstore.Logstore {
+	return n.store
 }
 
-func (t *net) GetHostID(_ context.Context) (peer.ID, error) {
-	return t.host.ID(), nil
+func (n *net) GetHostID(_ context.Context) (peer.ID, error) {
+	return n.host.ID(), nil
 }
 
-func (t *net) CreateThread(_ context.Context, creds thread.Credentials, opts ...core.KeyOption) (info thread.Info, err error) {
-	id := creds.ThreadID()
-	if err = t.ensureUnique(id); err != nil {
+func (n *net) GetToken(_ context.Context, identity thread.Identity) (tok thread.Token, err error) {
+	msg := make([]byte, tokenChallengeBytes)
+	if _, err = rand.Read(msg); err != nil {
+		return
+	}
+	sig, err := identity.Sign(msg)
+	if err != nil {
 		return
 	}
+	key := identity.GetPublic()
+	if ok, err := key.Verify(msg, sig); !ok || err != nil {
+		return tok, fmt.Errorf("bad signature")
+	}
+	return thread.NewToken(n.getPrivKey(), key)
+}
 
-	args := &core.KeyOptions{}
+func (n *net) CreateThread(_ context.Context, id thread.ID, opts ...core.NewThreadOption) (info thread.Info, err error) {
+	args := &core.NewThreadOptions{}
 	for _, opt := range opts {
 		opt(args)
 	}
+	// @todo: Check identity key against ACL.
+	identity, err := thread.ValidateToken(n.getPrivKey(), args.Token)
```

Comment:
Validate the token (if present). This is where the ACL will get checked.

Reply:
I'll take advantage of making a future-related question.

The ACL covers constraints on two dimensions, read and write.
Sounds to me that the _auth_ scheme of this _token_ generation etc, is to cover more the _read_ side of the ACL. So, to answer the following questions:
- I received a request, is this identity known in the ACL? If not, kick out. (So, just general identity check)
- OK, this identity is at least known. He is trying to pull the thread... is he allowed to? (So, read-related actions)

So now comes the important part of the question (at least for me and my understanding, I think @carsonfarmer  was interested in this topic too):
The _write_ side of ACL checks are tied to the _AutorSig_ at the record-level, right?
This because something you insisted about (which is important and true), that received records to be added to a log could be received by another peer than the author. This means that write-related ACL verifications can't be done at the _auth/token_ "layer"... so not here in this `ValidateToken`. Correct?

So, I'd like to confirm with you @sanderpick if that's the same mental model you have in mind.

I've other questions about _AuthSig_ at the record level (as I kind of mentioned in some previous review), but better leaving that for later.

Reply:
> The write side of ACL checks are tied to the AuthorSig at the record-level, right?

Yep, that's right. But the `core.Net` level _can_ do some write validation in `CreateRecord`/`AddRecord`. But this isn't enough on its own—you may have taken an existing thread that you don't have write permission in, hacked your node at the `core.Net` level to ignore outbound write checks (in `CreateRecord` or `AddRecord`), and then added a replicator. Now you send me an invite pointing to your replicator, which starts pushing me records. Based on the thread ID and `AuthorSig` I can verify that in fact you do not have write permission and I can ignore your records.

Reply:
One last point here, _read_ permission at the lower net p2p level is currently only handled by the _service key_, since reads come from a peer, not an identity. So, if you have the service key, you will be able to get logs and records. There are two problems with this:
1. Whether or not somebody is allowed to have the read key is not baked into the thread ACL rules.
2. I _think_ a peer could trick another peer into delivering them a service key by somehow getting them to use an invite that points back to their host address. However, this would require the invite user to already know the service key. So, not sure if this is a real problem. 

Consider these open questions!

Reply:
(edited above... I had mentioned a read key but that should have been service)

Reply:
Hmm, _whether or not somebody is allowed to have the service key_ could be baked into the ACL by using the remote's private key as identity. But still, a node could be hacked to ignore this _outbound_ rule. The ACL rules really only work on the receiver side.

Reply:
Considering the above comments, and the other at the PR level... I'd consider we're on the same page... so sounds good!

Now the other big question: _AuthSig_. Trying to explain my question might be long, so I'll take another approach and propose something and see your response to it. I hope that can answer many questions that would be long to do. :P

What if when a log is created, the first record is the public key of the owner (a pubkey in the ACL). This would mean that the log is saying that any record signed by the log keys are also indirectly signed by the author. (Because when the log was created, it said: _The author of this log is X_, and logs are single-writers obviously). Would that allow to avoid using the _AuthSig_ field in *all* the records of the log, and only leaving the _LogSig_ as is now?

Reply:
Hmm, clever! That's much cleaner. I like it 💯 

Let me take a quick look, maybe I can tag that on quickly. It would be nice to not have `AuthorSig` in there ASAP if we don't need it.

Reply:
So, when the ACL is in place, the _receiver of a log_ would have to then unpack the first record and verify that the author has write access (check ACL write rule)? If that's what you mean, wouldn't that imply that the receiver _has to have the read key_ in order to do the unpacking?

Reply:
First question: yes.

Second question: Uhm, no?. Isn't the read key needed for the Events? These signatures are on the Record. Am I confused?

Reply:
Oh, I thought you meant write a record that has a body (event) that contains the author's public key. Do you mean, keep `AuthorSig` on record (so readable by service key), but only on the very first record? Or remove `AuthorSig` and for the very first record, use the author's key to create `Sig`, instead of the log key?

Reply:
Well, actually didn't thought about a particular implementation.

Both sounds reasonable. One looks "cleaner" than the other, but it has a field only used once.
What are your thoughts? Not a clear-cut winner for me.

Reply:
cc @carsonfarmer since we're going down the path for an ACL related thing.

Reply:
As an aside, I realized here just having the author sig is not enough, we also need the pubkey. So, I'm glad we're having this convo ;)

Okay cool. Ideally, even the first record has the log signature as well. How about this: 

```
type record struct {
	Block     cid.Cid
	Sig       []byte
	AuthorKey []byte
	Prev      cid.Cid `refmt:",omitempty"`
}
```

This may be a little wild, but since we can control the key type of the logs, we could use the asymmetric encryption methods we already have in `crypto` to _encrypt the author sig in `Sig`_ for the first record. On the other end, verification would mean _decrypting `Sig` with log private key, and then verifying the author using the result as a signature. 

Alternatively, we just include a second signature in the first record :)

Reply:
I think I understood if I only changed _with log private key_ with _with log public key_, was that the intention? So you're leaning towards "reusing" the field and making the first record a special case. Right?

Reply:
I'm finding it really hard to track this conversation, but I thinks its making sense to me. My only concern with making the 'first' record special is that it means a client can't really trust or do anything until they have synced a full log, right? So assuming they come in 'late' to the game, this could be an issue, yes?

Reply:
Uhm, and what's `AuthorKey` used for in that idea?

Reply:
> I'm finding it really hard to track this conversation, but I thinks its making sense to me. My only concern with making the 'first' record special is that it means a client can't really trust or do anything until they have synced a full log, right? So assuming they come in 'late' to the game, this could be an issue, yes?

Yes, that's true. Seems like a tradeoff between covering that case, and having some overhead in all the headers of the log. Maybe we can include the info every 200 records or have some middle ground? Just a flashing thought.

Reply:
Ah, scratch that encryption idea... I did mean private logs key, but I forgot (duh) that the receivers won't (necessarily) have the private key for decryption. 

`AuthorKey` is used to indicate to the receiver who the author is. If they just had a signature, you wouldn't be able to tell who it is or verify.

Reply:
Hrm, that's a good point @carsonfarmer. Though, every record _could_ carry a link to the first.

Reply:
The first record could have:
- `pubkey`, a new field which should map to some pubkey in the ACL. Indicates author pub key. Only has value in first record.
- `Sig` field would be the signature of that pubkey of the record. So this is the only case `Sig` isn't a signature made with the log-private-key.
- No `AuthorSig` field.

What I'm missing in this reasoning?

Reply:
Funny, I was just typing out... "the sig could just be the sig of the author's pubkey", but now I see that's what you meant :P

Okay, so I'll change the logic and `AuthorKey` -> `PubKey`.

Reply:
Nice, I like it. I updated the signing+verification to take the first record into account. For now, I'm just including the pubkey in every record, even though only the first one is verifiable. Once we jump on the ACL and the _log record height_ issue (#137), we can adds these links... sound okay? I've already ballooned this PR a good bit.

Reply:
SGTM!